h/sql_tasks/sql_file.py

```diff
@@ -0,0 +1,67 @@
+import os
+import os.path
+from dataclasses import dataclass
+from typing import List
+
+import jinja2
+
+from h.sql_tasks.sql_query import SQLQuery
+
+
+@dataclass
+class SQLFile:
+    """A class representing an SQL file with multiple queries."""
+
+    path: str
+    """Full path of the script."""
+
+    environment: dict
+    """Template vars to pass to templated SQL statements."""
+
+    queries: List[SQLQuery] = None
+    """Queries contained in this file."""
+
+    _jinja_env = jinja2.Environment()
+
+    def __post_init__(self):
+        with open(self.path, encoding="utf-8") as handle:
+            script_text = handle.read()
+
+        if self.path.endswith("jinja2.sql"):
+            # Looks like this file has been templated
+            script_text = self._jinja_env.from_string(script_text).render(
+                self.environment
+            )
+
+        self.queries = [
+            SQLQuery(text=query, index=index)
+            for index, query in (enumerate(self._parse_multi_query_sql(script_text)))
+        ]
+
+    @classmethod
+    def from_dir(cls, task_dir: str, environment: dict):
+        """
+        Generate `SQLFile` objects from files found in a directory.
+
+        This will return a generator of `SQLFile` based on the natural sorting
+        order of files found in the directory, and subdirectories. Only files
+        with a `.sql` prefix are considered. Files with `.jinja2.sql` are
+        treated as Jinja2 templated SQL and are rendered using the provided
+        environment.
+        """
+        if not os.path.isdir(task_dir):
+            raise NotADirectoryError(f"Cannot find the task directory: '{task_dir}'")
+
+        for item in sorted(os.listdir(task_dir)):
+            full_name = os.path.join(task_dir, item)
+
+            if os.path.isdir(full_name):
+                yield from cls.from_dir(full_name, environment=environment)
+
+            elif full_name.endswith(".sql"):
+                yield SQLFile(full_name, environment=environment)
+
+    @staticmethod
+    def _parse_multi_query_sql(script_text):
+        # It's an understatement to suggest this could be better...
```

Comment:
We could evaluate https://github.com/andialbrecht/sqlparse which seems to also support formatting which is something we could add to our `make format`.

Demo of the formatting here: https://sqlformat.org/


If we don't want to introduce a dependency just yet we could add a `and not query.startswith("--")` to remove that gotcha in the comments.